### ISSUE TITLE
Change log rotation scheme, again

### ISSUE THREAD

**Issue by sc-perth** [certbot/certbot#3602]
Certbot's built in log rotation prevents logs from being handled in the manner defined by the system administrator. I see that in #3382 ([2bdf125](https://github.com/certbot/certbot/commit/2bdf1258ad9db1a3a0681a3b81614c93a1d5b96d)) we discovered one aspect of this issue and addressed that by increasing the number of logs kept by Certbot to 1000.

I object to this fix for a few reasons:
1) Systems not specifically configured to address Certbot's logs will end up with 1001 log files.
2) It is not typically in the utility's scope to address log rotation.
3) An administrator seeking to control Certbot's logging must now consider how to work around the existing log rotation, instead of simply addressing log size & count in the normal fashion.

For these reasons, please consider removing the built in (and hard coded) log rotation. That will bring Certbot's log file handling into line with other *nix utilities. If it remains necessary to call logger.handlers.RotatingFileHandler; please then set the size & count to 0; this will cause it to constantly write to the same file. 

Another note is that some/all log rotators can inform a process when a log is rotated; it may be desirable to enable this integration. Example: `path_to_pid_cmd_file` section in the [newsyslog.conf man page](http://www.unix.com/man-page/freebsd/5/newsyslog.conf/) & `postrotate` section in the [logrotate man page](https://linux.die.net/man/8/logrotate).

Lastly, adding a section to the documentation on Certbot's logging may be advised.

Thank you!

**Comment by pde** [comment]
@jsha thoughts? Interested in taking this?

**Comment by jsha** [comment]
I agree the current hard-coded log rotation is bad. I increased the number of logs to avoid the most pernicious problem, that logrotate + certbot rotation silently drops logs. That problem is still present, but only if logrotate is configured to retain more than 1000 files, which is rare.

I'm afraid I don't have the time for the more thorough fix, but I would encourage anyone to take it!

It's also worth thinking about how to configure "out of the box." The Debian / etc packages can install something in logrotate.d. However, if someone installs with certbot-auto on a machine with low disk, it's likely that we would eventually cause an out of disk error unless they do additional configuration. Maybe we just decide that our daily log growth, when run daily, is small enough not to worry about it?

**Comment by sc-perth** [comment]
I have 2 questions that I think will help this move forward.
1. @jsha Will _certbot_ have issues if it simply constantly appends to the same log file (disable all built in log rotation)?
2. What were the design considerations that caused log rotation to be implemented in the first place?

**Comment by jsha** [comment]
1. Probably not. When asking people to provide log files to help debug a problem, that will result in much bigger uploads. But it will also help provide more context on what they tried, so might be a win there.
2. I don't know, I wasn't part of reviewing that change. My guess would be: We would like our out-of-the-box configuration to be able to renew certificates indefinitely without causing a box to fail because it ran out of disk space.

**Comment by sc-perth** [comment]
So it seems there is no _need_ for certbot to handle logging in the current manner. You do however bring up a good reason we can't just constantly log to the same file though, certbot-auto. Since that is supported we can't make assumptions that package maintainers have seen to dealing with the log file. Also as you say, we don't want to take over someone's disk space, which demands log rotation in some manner.

So the challenge is how do we get certbot to create logs that will be handled "correctly" under both supported installation types. I believe the answer is to have certbot take advantage of Python's syslog integrations. Since syslog() is part of the POSIX specification for system interfaces, this should provide an interface though which logging can be accomplished "correctly" on most systems, no matter the installation method. I would appreciate feedback on this conclusion.

I'm going to try switch to [SysLogHandler()](https://docs.python.org/2/library/logging.handlers.html#sysloghandler) on a local repo. It will be a learning experience for me, any input would be appreciated. Especially which logging facility to go through. For now I'm selecting daemon, it's a bit off, but not really wrong either. Decent reference: [Ubuntu's LinuxLogFiles page](https://help.ubuntu.com/community/LinuxLogFiles).

**Comment by sc-perth** [comment]
Alright, I think I've figured it out. Please review [my proof of concept python script](https://github.com/sc-perth/certbot/blob/sc-perth/POSIX-logging-test.py). It should make the path I'm planning on taking here pretty clear. It may also make clear just how new to this I am. Please also pay special attention to the find_facility() function. I have no clue what including that would mean from a legal point of view. It can honestly be yanked, but not if we want to make the syslog facility something that can be configured. I could write one up from scratch, though it would look quite similar.

In `certbot/main.py` I'm confused by the presence of a root_logger, which is created at [L641](https://github.com/certbot/certbot/blob/master/certbot/main.py#L641). It does not appear to be used, see [this search](https://github.com/certbot/certbot/search?utf8=%E2%9C%93&q=root_logger). It appears we are only using the logger created at [L44](https://github.com/certbot/certbot/blob/master/certbot/main.py#L44) ([search](https://github.com/certbot/certbot/search?utf8=%E2%9C%93&q=logger&type=Code)). Let me know if there's some reason I shouldn't get rid of the root_logger while I'm mucking about.

**Comment by jsha** [comment]
I like the idea of logging to syslog by default, but I think it creates a problem: We can't tell people "send us a copy of your /var/log/letsencrypt.log," because the logs will wind up stored in different places depending on the local syslog config, and are most likely interleaved among other, non-Certbot logs (some of which may be sensitive).

It does have the advantage that it's one less directory we need to try and create on startup. But I think the inability to give people a single path where they can definitely find their log is a serious problem - do you have an idea how to solve it?

**Comment by sc-perth** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

You make a valid point. It should be considered carefully. Unfortunately I'm going to bed and this will be a busy week too.  So for now the short answer is: ```grep certbot /var/log/daemon.log >> temp```.

I intend to think about this more carefully, later though.
- --
Between A & B there's C.
GPG: 0xe818cb04b0a635fc
5b55 a483 138e 69e8 7be1
6bbe e818 cb04 b0a6 35fc
https://sks-keyservers.net/pks/lookup?op=get&search=0xe818cb04b0a635fc
Sent from my phone usind K9 Mail, pardon the typos (I hate autocorrect).

On October 24, 2016 8:25:51 PM EDT, Jacob Hoffman-Andrews <notifications@github.com> wrote:
>I like the idea of logging to syslog by default, but I think it creates
>a problem: We can't tell people "send us a copy of your
>/var/log/letsencrypt.log," because the logs will wind up stored in
>different places depending on the local syslog config, and are most
>likely interleaved among other, non-Certbot logs (some of which may be
>sensitive).
>
>It does have the advantage that it's one less directory we need to try
>and create on startup. But I think the inability to give people a
>single path where they can definitely find their log is a serious
>problem - do you have an idea how to solve it?
>
>--
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/certbot/certbot/issues/3602#issuecomment-255903003
-----BEGIN PGP SIGNATURE-----

iQFDBAEBCgAtJhxKb2huIEQuIEhhcmxhbiA8cGVydGhAc2VtaWNvc21pYy5uZXQ+
BQJYDr80AAoJELHPXx+MTRDUz4MH/jJ42/R02CmhSKhr/ntoGhxUsiz3onQn8hWJ
FP6Eb9TV3NzrEuQHVRbr69LatiT3VdXE3MKFCVJTxNCCUSCwA/LEcEiOKIcq/7PW
xhvyHFfg3T5ZwyiGUOaQ4yVqodWHLlXAMvu++VBR1wgLJU7T2SNmxt+0vO+GxdNS
FDQwtH8sn73p4LnIR+cPtPRKk7HZ0SiKOHNyDtXpaRnPcBdWExp25nEOGpcK/oij
wW1Zh2SRGyC9g2g/sM8Zw2WS1KK69KpuUASg4FQ0bXxSGQZTKRafERxmWG17Tiqv
fWD44c9aZNiRE3pUoQv0fRKmBTvb2X/cplOwC4YL4uJwog/eSq0=
=C4eu
-----END PGP SIGNATURE-----

**Comment by timwienk** [comment]
I see this issue has been dormant for a while. In case I missed something: is this still the issue to track for the logging situation?

Under the assumption that it is and that there has been no progress made in this area:

What about leaving the default logging configuration as it currently is - until a definitive decision has been made - and make the log rotation configurable? Even if it's nothing more than a flag to turn log rotation completely off, surely that's a safe enough feature that can retained in the future as well. This ensures two things:
1. with default configuration things work the way people are used to.
2. with the flag set, the user consciously changed the configuration and can be expected to know how to handle the both log rotation and reporting/providing log entries if necessary.

**Comment by sc-perth** [comment]
My life has been a bit crazy, thus the issue being dormant, but stabalized recently, see my profile for proof :). The flag seems like an entirely reasosable suggestion. I plan on giving this some attention tomorrow. Sorry for the delay.

(Then Christmas will make my life crazy again. Anyone is free to "steal" fixing this issue.)
-- 
Between A & B there's C.
GPG: 0xe818cb04b0a635fc
5b55 a483 138e 69e8 7be1
6bbe e818 cb04 b0a6 35fc
https://sks-keyservers.net/pks/lookup?op=get&search=0xe818cb04b0a635fc

**Comment by sc-perth** [comment]
Not the easiest decisions I've ever been asked to make... :)
Here's what I've just finished concluding might be the best path forward after poking about some more. Feedback is quite desirable.

### tl;dr:
--logs-dir will cause confusion and will warn of future deprecation; --log-target will replace; --logs-dir becomes an alias for --log-target. Constantly fallback to current method if there is any confusion. Couple of new exit points. --log-target will check if pointed at a syslog facility's file, if so go through syslog facility, else write to file indefinitely. Add a --rotate-logs flag for future use & people that don't want to configure their log rotator.

### The proposed new logic in all its spaghetti: 
--logs-dir needs deprecated, because it keeps causing confusion in every path forward I think of. Replace it with --log-target. Now I've created a huge backwards compatibility issue, so we'll have to keep --logs-dir around, warn people when they use it, but actually use --log-target.
* If --logs-dir is pointed at a directory, use old logging method & warn user of future deprecation.
* If --log-target is pointed at a directory, throw temper-tantrum :) & exit.
    * (*If exit codes have been standardized within certbot please inform me.* Otherwise I'll use 64, command line usage error, as specified in /usr/include/sysexits.h)
* --log-target/--logs-dir will check to see if it's being pointed at a _file_ in /var/log (not sub directories). If so it will check to see if that file corresponds to a valid syslog facility. If so logging will be directed through the chosen syslog facility.
    * (/var/log/syslog -> syslog's 'syslog' facility; /var/log/daemon -> syslog's 'daemon' facility; etc)
* If --log-target/--logs-dir has been pointed at a _file_ that is not a valid syslog facility (assumed to be the case if not a file in /var/log), then append to the end of the file indefinitely.
    * If target does not exist:
        * --log-target will conclude it was pointed at a file, create the file & append to it indefinitely.
            * (If someone is manually specifying a non-standard log file, it's their job as the owner of the system to manage that log file; this is a completely normal expectation provided the default doesn't create such a file w/out handling the situation properly).
        * --logs-dir will assume it was pointed at a directory, create the directory, and fallback to current method.
* Create a new --rotate-logs argument. This will enable continuing with the current method moving forward.
    * If --rotate-logs is specified in addition to --log-target being pointed at a valid syslog facility's log file, throw a helpful error about how certbot will not rotate system logs & exit.
* If --log-target/--logs-dir was not declared, fallback to current method.
    * Equivalent to `--log-target /var/log/certbot/certbot.log --rotate-logs` under proposed changes.
    * I have no objections to this being the default moving forward. It's easy enough to add 1 more argument to my certbot script.

**Please comment** & know this is going to take me a bit, consider the time of the year.

I'm also fairly certain that my syslog-ng.conf in borked: https://bpaste.net/show/3666e9b82078
So I would appreciate it if someone could run my [POSIX-logging-test.py](https://github.com/sc-perth/certbot/blob/sc-perth/POSIX-logging-test.py) `python2.7 ./POSIX-logging-test.py` and make the effects on their logs visible to me (script sleeps 1 sec between writes & 3 secs between facilities. Takes 2-3 mins). I'm using that data to validate how the syslog facilities correspond to specific log files. Plenty of methods exist to extract the log entries, all lines will contain 'sc-perth_log-test', but I'm going to recommend running my [logtail.sh](https://github.com/sc-perth/BashScripts/blob/master/logtail.sh) script and tee `./logtail.sh | tee -a ./sc-perth_log-test` (in a separate terminal). Upload/paste the file somewhere after you've checked it for sanitization. (Scripts are on GitHub).

**Comment by timwienk** [comment]
Firstly, thanks for putting the effort in right away. I hope you know
that I did not aim for you to throw your busy schedule around just to
dive back into this, however I do very much appreciate it being looked
into again.

I am rather busy myself right now (partially due to time of the year), I
can try and see if I can locally run your tests soon. I think everyone
will (or at least should) realise that a potentially breaking change
like this will probably not be completely finished, let alone released,
in december.

Your self-proclaimed logical spaghetti seems a rather complete solution,
but it does feel a bit complicated. Wouldn't a solution like the
following be easier and less intrusive?

Alternative solution I
----------------------

1. if no `--log-target` or `--logs-dir` specified:
    1. default to `--log-target /var/log/letsencrypt/` (or
       `/var/log/certbot/` if you prefer).
2. if `--logs-dir` is specified:
    1. deprecation warning,
    2. continue as if the value was specified for `--log-target` (maybe
       append `/`?).
3. if `--log-target` is specified (or obtained through `--logs-dir` as
   noted above):
    1. if `--log-target == "stderr"`, log to `stderr` (like the default
       action),
    2. if `--log-target == "syslog"`, require `--syslog-facility` to be
       specified and log to syslog,
    3. if `--log-target` points at a directory (maybe even require
       target ending in a `/`?), log to specified directory (create if
       necessary) and do rotation (old behaviour) unless
       `--no-log-rotation` is specified,
    4. else: log to the file specified (create if necessary) and don't
       rotate, since we only know about one file.

Alternative first step to alternative solution I
------------------------------------------------

Which I would personally prefer, but has more effect on current
behaviour:

1. if no `--log-target` or `--logs-dir` specified:
    1. warning: `"default log target will change to stderr in the next
       version"`,
    2. continue as if `--log-target /var/log/letsencrypt/` was specified
       (old location + behaviour),
    3. in the future: default to `--log-target stderr`.

This last idea is obviously (even more) breaking, and you guys may not
agree since it also may make it harder for people to find the right data
to provide in case necessary. The idea behind this is that `certbot` by
default is a command line tool, which makes logging to a file by default
illogical to me. Default logging to a file is something I'd expect of
automated tools (and for that purpose it can be configured - but even
then things like `cron` handle `stderr` output fine) or
daemons/services.

**Comment by timwienk** [comment]
@sc-perth, I ran your logging test script with your log-tail script. Fresh Debian Jessie installation, rsyslog 8.4.2 (with Debian's default configuration). Hope it helps.

https://gist.github.com/timwienk/d1eef0c0553d63683f67387e0066c615

**Comment by nyetwurk** [comment]
Is there any progress here? Or should I just write a cronjob to delete old logs?

**Comment by sc-perth** [comment]
There's not really been any progress. Every time I come look at this I get confused, frustrated, and/or tired. But I was looking through this again yesterday and paid more attention to a comment @timwienk made.
> [... Even if it's nothing more than a flag to turn log rotation completely off, surely that's a safe enough feature that it can be retained in the future as well. ...](https://github.com/certbot/certbot/issues/3602#issuecomment-264656681)

So here's a patch that adds the --disable-log-rotation option.

```
diff --git a/certbot/cli.py b/certbot/cli.py
index 8acb1f6..94e55c4 100644
--- a/certbot/cli.py
+++ b/certbot/cli.py
@@ -668,6 +668,12 @@ def prepare_and_parse_args(plugins, args, detect_defaults=False):  # pylint: dis
         None, "-t", "--text", dest="text_mode", action="store_true",
         help=argparse.SUPPRESS)
     helpful.add(
+        [None, "automation", "paths"],
+        "--disable-log-rotation", dest="disable_log_rotation",
+        action="store_true", default=False,
+        help="Disable per-run rotation of the log. This results in a single "
+        "log file.")
+    helpful.add(
         [None, "automation"], "-n", "--non-interactive", "--noninteractive",
         dest="noninteractive_mode", action="store_true",
         help="Run without ever asking for user input. This may require "
diff --git a/certbot/main.py b/certbot/main.py
index aa9b189..230210d 100644
--- a/certbot/main.py
+++ b/certbot/main.py
@@ -597,8 +597,12 @@ def setup_log_file_handler(config, logfile, fmt):
     """Setup file debug logging."""
     log_file_path = os.path.join(config.logs_dir, logfile)
     try:
-        handler = logging.handlers.RotatingFileHandler(
-            log_file_path, maxBytes=2 ** 20, backupCount=1000)
+        if config.disable_log_rotation:
+            handler = logging.handlers.RotatingFileHandler(
+                log_file_path, maxBytes=2 ** 20, backupCount=0)
+        else:
+            handler = logging.handlers.RotatingFileHandler(
+                log_file_path, maxBytes=2 ** 20, backupCount=1000)
     except IOError as error:
         raise errors.Error(_PERM_ERR_FMT.format(error))
     # rotate on each invocation, rollover only possible when maxBytes
```

**Comment by nyetwurk** [comment]
@sc-perth my preference would actually be to add a backupCount option, since by default most distros (e.g. debian) do not provide a logrotation config file for certbot anyway.

**Comment by nyetwurk** [comment]
so maybe -log-rotation-max which, if 0, disables rotation, or overrides the "1000" default?

**Comment by sc-perth** [comment]
Bluntly, I'm not interested in that. _I_ don't want certbot handling log rotation. _I_ want certbot going through syslog. That said, for situations where certbot _is_ handling log rotation, that's a darn good idea. And it's trivial enough to implement, now that I've got the argument handling figured out. So... *in a feigned put-upon tone" I _guess_ I can do that _too_.

**Comment by sc-perth** [comment]
Done?
Commit: https://github.com/sc-perth/certbot/commit/3d55a473eff7538aecdc9351c35b8fb5600e862b
Compare against certbot/certbot/master: https://github.com/certbot/certbot/compare/master...sc-perth:3602-disable-log-rotation
Patch against same point: https://github.com/certbot/certbot/compare/master...sc-perth:3602-disable-log-rotation.patch

**Comment by nyetwurk** [comment]
@sc-perth much appreciated. I will take a look time permitting.

**Comment by nyetwurk** [comment]
@sc-perth I should add I agree with you 100% that certbot should (generally) not be doing any log rotation either, and it should be (generally) using syslog.

**Comment by sc-perth** [comment]
OK. Passes testing for me now.
Commit: https://github.com/sc-perth/certbot/commit/40dee6a8ad6191380b2af4ef02c7286ca4c189de
Compare against certbot/certbot/master: https://github.com/certbot/certbot/compare/master...sc-perth:3602-disable-log-rotation
Patch against same [here](https://github.com/certbot/certbot/compare/master...sc-perth:3602-disable-log-rotation.patch)

**Comment by nyetwurk** [comment]
@sc-perth - works fine for me as well. Stupid idiot digression question: can that option be put in the /etc/letsencrypt/renewal config files or does it have to be on the command line every time?

**Comment by sc-perth** [comment]
:+1: Good question! I'm going to assume it's CLI only as I haven't gone out of my way to do anything with/for a config file. That said, with how all of this is utilizing certbot code it's entirely possible that integration has been handled. Why not try adding it to a test config file, specify the config file with `--config CONFIG_FILE` on a `--dry-run` and let me know? I'd appreciate that, I haven't ever used the config file since I custom script the whole certbot proces :)

**Comment by St-Ranger** [comment]
@sc-perth 
I saw the pull request #4195, and I thought I'd add a few comments/suggestions for the implementation (in part related to the suggestion made by @bmw )

**1.** As a user, I would easily understand that `--disable-log-rotation` disables log rotation, thus allowing it to grow infinitely (in the absence of other log rotators).
In contrast, the meaning of  `--max-log-count` set to 1 can be ambiguous. Besides what @bmw suggested, it can also mean (and that would be _my first thought_): rotate the log, but keep only one file. And then there is yet another ambiguity (related to "off-by-one): keep one old file or one file total?

Logical of at least some of the syslog rotators is that you configure the number of "_rotated_" (aka old) logs. E.g. FreeBSD newsyslog(8) uses an option `count`: "_count denotes how many rotated log files should be kept._ " See "Log Management and Rotation" with the example of newsyslog.conf(5) here:
https://www.freebsd.org/doc/handbook/configtuning-syslog.html

So, my suggestion would be:

- [ ]  keep `--disable-log-rotation` as the **only** way to disable rotation. 
- [ ] Rename  `--max-log-count` to e.g. `--max-old-log-count.` 
With that option set to 0 **and without** `--disable-log-rotation`, it would mean: rotate and discard all old logs (keep the latest one only).

I don't think this would require too many changes of the code, but this would make it much less ambiguous to the user (and hopefully even unambiguous).

**2.** While we are the topic of log rotation, let me share my initial experience about it. I was rather surprised when I saw multiple log files from one day, and some of them were very small. I ultimately understood that each file corresponded to each certbot command issued. E.g. "certbot certificates" generates a logfile that is only 995 bytes long. I understand that it is convenient from the point of view of debugging certbot. It is not very practical from the point of view of a production system.  So, my first inclination was to propose a `--max-log-size` option that would determine when the logs should be rotated.
I can see that the upcoming --disable-log-rotation in combination with the use of the system-based log rotators mitigates this situation. So, that seems to be alright.

**3** Wish-list: it would be nice to be able to configure the **level** of logging, especially when --disable-log-rotation is implemented. I thought it was possible, but I am not seeing it among the options (sorry, I didn't look into the source code): https://certbot.eff.org/docs/using.html . Am I missing it somewhere?

**4** Finally, regarding the question from @sc-perth about necessity of syslog alternative. I think it would be very useful and important for some very frequent use cases. Let me give just two reasons.
a) syslog can do more than just storing to a local file. It can do remote logging, it can throw messages to the screen of specific users or to  the console (does anybody have/use a console anymore? ;-)  ) ;
b) it allows separating messages with different levels (crit, error, info, debug, ...)
Many of those functions could be important for people who have multiple servers deployed and need to be alerted about, say, a problem with the cert renewal.
Those syslog functionalities exists on (all known to me flavors of) Unix, so that people do not invent new log analysis/management for each new application.
I am sure that you know all of this, but I thought I'd remind about it.

So, in conclusion, I think there are use cases that would benefit from certbot's own log files, and those that would benefit from the integration with syslog. So, it makes perfect sense to continue that integration. "May the Force be with you!" :-)

**Comment by bmw** [comment]
Fixed in #4907.
